packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -3504,6 +3084,282 @@ merge_matrices(CrsMatrixStruct<Scalar, LocalOrdinal, GlobalOrdinal, Node>& Aview
+    {}
+
+    KOKKOS_INLINE_FUNCTION void operator()(const GO i) const
+    {
+      lids(i) = localMap.getLocalElement(gids(i));
+    }
+
+    GView gids;
+    LView lids;
+    LocalMapType localMap;
+  };
+} //end ColMapFunctors
+
+//Build the minimal (sorted) column map for the given set of global columns
+//Then convert gids and store them in lids (gids is not modified)
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+Teuchos::RCP<Map<LocalOrdinal, GlobalOrdinal, Node> > AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+makeColMapAndConvertGids(GlobalOrdinal indexBase, GlobalOrdinal minCol, GlobalOrdinal maxCol,
+                   const typename MMdetails::AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::global_col_inds_array& gids,
+                   typename MMdetails::AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::col_inds_array& lids,
+                   const Teuchos::RCP<const Teuchos::Comm<int>>& comm)
+{
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+  int myRank = comm->getRank();
+  std::cout << "Rank " << myRank << ": column map for some subset of global cols: " << minCol << " <= col <= " << maxCol << '\n';
+  typedef typename MMdetails::AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::global_col_inds_array::non_const_type GView;
+  typedef typename MMdetails::AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::col_inds_array::non_const_type LView;
+  typedef typename Map<LocalOrdinal, GlobalOrdinal, Node>::local_map_type LocalMap;
+  GlobalOrdinal nentries = gids.extent(0);
+  typedef Kokkos::Bitset<device_type> bitset_t;
+  typedef Kokkos::ConstBitset<device_type> const_bitset_t;
+  //Determine the set of GIDs in the column map using a dense bitset (covers entire possible range)
+  bitset_t presentGIDs(maxCol - minCol + 1);
+  Kokkos::parallel_for(range_type(0, nentries), ColMapFunctors::GatherPresentEntries<GlobalOrdinal, GView, bitset_t>(minCol, gids, presentGIDs));
+  execution_space().fence();
+  const_bitset_t constPresentGIDs(presentGIDs);
+  GView mapGIDs("Sum colmap GIDs", constPresentGIDs.count());
+  std::cout << "Rank " << myRank << ": determined there are " << mapGIDs.extent(0) << " unique global cols on this proc.\n";
+  Kokkos::parallel_scan(range_type(0, maxCol - minCol + 1), ColMapFunctors::ListGIDs<GlobalOrdinal, GView, const_bitset_t>(minCol, mapGIDs, constPresentGIDs));
+  execution_space().fence();
+  //Construct the map
+  auto colMap = rcp(new Map<LocalOrdinal, GlobalOrdinal, Node>(
+        Teuchos::OrdinalTraits<Tpetra::global_size_t>::invalid(), //have Tpetra compute global #GIDs
+        mapGIDs, indexBase, comm));
+  //Get the device-friendly local map
+  LocalMap localColMap = colMap->getLocalMap();
+  //Use it to convert gids to lids
+  Kokkos::parallel_for(range_type(0, nentries), ColMapFunctors::ConvertGlobalToLocal<GlobalOrdinal, GView, LView, LocalMap>(gids, lids, localColMap));
+  execution_space().fence();
```

Comment:
We shouldn't sprinkle fences around without understanding why -- I don't think we need it here.  See above comment about adding a fence to Map's constructor.

Reply:
colMap is fully constructed by now, but this parallel_for (ConvertGlobalToLocal) populates lids, which is the other output of this function. Shouldn't we make sure it can be immediately accessed on device by any caller? In theory, if the parallel_for took a long time, having lids be partially computed could cause problems in the expertStaticFillComplete.

Reply:
@brian-kelley It populates LIDs on device, so any subsequent device function on the same CUDA stream should be able to access the LIDs without an intervening fence.  Device kernels on the same stream are sequenced.

Reply:
Trilinos developers would like to remove both the `CUDA_LAUNCH_BLOCKING` and UVM assumptions at some point.  Right now, the fences are just extra work, because Trilinos requires `CUDA_LAUNCH_BLOCKING` to be set.  Removing the `CUDA_LAUNCH_BLOCKING` requirement will take lots of work to find existing issues.  I'd rather not hide any existing correctness issues by doing unnecessary fences.  It's like doing extra `MPI_Barrier`s to hide MPI correctness issues -- it doesn't necessarily make the code correct.